### ISSUE TITLE
Add new dismissible CTA widget in Monetization section for when AdSense is inactive

### ISSUE THREAD

**Issue by tofumatt** [google/site-kit-wp#4145]
## Feature Description

The Monetization section of the Unified Dashboard should include a dismissible "Connect to AdSense" widget that displays when AdSense is not connected. If AdSense is connected, the widget should return `null`.

<img width="908" alt="Screenshot 2021-09-28 at 15 40 25" src="https://user-images.githubusercontent.com/90871/135109497-26669e2b-1480-4d4f-9a53-1db7a3c7a8bc.png">

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* There should be a new widget created using a new component (`AdSenseConnectCTA` or similarly named) that is registered to the `AREA_MAIN_DASHBOARD_MONETIZATION_PRIMARY` widget area when the `unifiedDashboard` feature flag is enabled.
* This widget should `return null` and output nothing when the AdSense module is connected.
* The widget should render the AdSense CTA widget as mocked up in this Figma file: https://www.figma.com/file/VILRieNSLg2DOVyEgFBkXe/%5BMVP%5D-Generic-Dashboard?node-id=2%3A10104
* The widget should be dismissible, which should set a permanent, per-user flag that this AdSense CTA has been dismissed. Once this widget is dismissed and this setting is set, the widget should also `return null`. (This will cause the entire area to be marked as "empty" after #4136 is ready.)
* The Figma mockup shows AdSense being connected in the page without a redirect, but for now a full-page reload to the "Setup" page for the AdSense module is fine.
* If AdSense is activated but not connected (eg setup has not completed), the widget should show a similar UI to the Figma mockup, but the copy should encourage users to complete set up of AdSense.
* The AdSense link about account review should point to https://support.google.com/adsense/answer/9724.

## Implementation Brief

Create a new  `AdSenseConnectCTA` component inside `assets/js/modules/adsense/components/common`.

* It should be designed according to the Figma file.
    * In the case of an active-but-not-fully-connected AdSense module state, change the "Connect" button to something like "Complete setup". Otherwise the module can stay the same for now.
* The Component is visually similar to `SetupMain` component of `adsense` and `idea-hub` modules, for layout and other visual stuff, follow them.
* Create a handler for `Connect` button:
    * It should have similar logic to `handleSetupModule` function of `assets/js/components/ModulesList.js` component. 
    * https://github.com/google/site-kit-wp/blob/4fcb58149ce46d346bed4ced38124d13bfe92025/assets/js/components/ModulesList.js#L50-L71
* Create a handler for `Not Interested` button:
* It will flip the state of `dialogActive` to `true`, thus opening the `Dialog` inside a `Portal`. (Mentioned in details below). 
* Add a handler for the `confirm` button; It is a async function that will do the following:
    * set `inProgress` state of the dialog to `true`.
    * await the `dismissItem` call (from `CORE_USER` data store action) with the parameter `adsense-connect-cta`.
    * set the `inProgress` and `dialogActive` false.
* Add the `Dialog` inside a `Portal` in the render section of the component:
    * Pass the aforementioned `dialogActive`, `inProgress`, `handleConfirm` etc props and title and subtitle from the AC/Figma design.
    * See [here](https://github.com/google/site-kit-wp/blob/4fcb58149ce46d346bed4ced38124d13bfe92025/assets/js/components/ResetButton.js#L129) for a example usage of dialog. 

Create a new  `AdSenseConnectCTAWidget` component inside `assets/js/modules/adsense/components/dashboard`.

* The component will be used as a `Widget` thus taking it as a prop.
* Determine whether the `adsense` module is connected by using the `isModuleConnected` from `CORE_MODULES` data store. 
* Similarly, determine whether the Adsense Connect CTA was dismissed by using the `isItemDismissed` selector from the `CORE_USER` data store, passing the aforementioned `adsense-connect-cta` as the slug. 
* if either ` moduleConnected` or `adsense-connect-cta` is not `false`, return `null` from the component. ie. 
```js
        if ( moduleConnected !== false || adsenseCTADismissed !== false ) {
            return null;
        }
```
* Otherwise, return the newly created `AdSenseConnectCTA` wrapped by the Widget prop. 

In `assets/js/modules/adsense/index.js`,

* Register the newly created `AdSenseConnectCTA` widget in the `AREA_MAIN_DASHBOARD_MONETIZATION_PRIMARY` widget area. 
    * It should be full width and lower priority than `adBlockerWarning`. 

### Test Coverage

* Add a test to confirm that the dismiss button dispatches the correct event/API request.
* Add a test to ensure the widget returns `null` when AdSense is connected or the item is marked as dismissed.

## QA Brief
* Disconnect AdSense module.
* Enable the `unifiedDashboard` feature flag.
* Go to the main dashboard where the widget will be displayed, prompting the user to connect `AdSense`
* Clicking the "Connect" button should start the flow to connect AdSense.
* Abort AdSense setup and return to the main dashboard.
* The CTA should now be "Complete set up" and clicking on it will continue setting up AdSense
* Clicking the "Not Interested" link will trigger a modal where confirmation will be needed to hide the widget.
* Check comment here: https://github.com/google/site-kit-wp/issues/4145#issuecomment-965098899 about visual design differences.

## Changelog entry

* Add dismissible AdSense CTA widget for monetization section in the unified dashboard.

**Comment by kuasha420** [comment]
Hi @tofumatt (since you've written the AC),

Some observations about the AC,

* I think `AdSenseConnectCTA` is a more appropriate name for the widget so I have used that in the IB, what do you think? 
* Is the usage of `dismissItem` and `isItemDismissed` from `CORE_USER` to set the dismissed user flag appropriate? 
* Also, while isDismissed/IsConnected is undefined, we should still return `null` from the component to avoid flashes, is that correct? 
* Lastly, we are not handling the Post-authenticate part of the `adsense` module activation in this issue, right? I think there is no infrastructure in place to handle that for any of the modules in Unified Dashboard yet, is that correct? 

Sending the IB for review, feedback is much appreciated!

Cheers.

**Comment by tofumatt** [comment]
Hi @kuasha420, great questions, thanks!

The answer to the first three is: yes! Your component name is much nicer, agreed. I've updated the ACs. Using `dismissItem` is right for this. And preventing the flash by returning `null` makes sense.

> Lastly, we are not handling the Post-authenticate part of the `adsense`

I think you mean that we aren't worried about what happens after you click "Connect AdSense" or whatever the CTA button/link is, right? If so: yes, that's correct. In this issue we're just worried about showing the CTA when relevant and having it go to our setup URL. We'll deal with actual module activation in another issue 👍🏻

**Comment by tofumatt** [comment]
I added a few test criteria and a point about the module being active but not connected to the ACs/IBs because I left them out originally and just realised now they're relevant. But this looks good to go 👍🏻 

IB ✅

**Comment by asvinb** [comment]
@tofumatt @kuasha420 I have some notes after I started working on this issue and wanted to clarify since it's not in the IB or AC:

-  In the Figma designs, the CTAs are normal case ("Connect", "Not interested", "Complete setup") however I remember there were some conversations about them being uppercase like we have right now. To be consistent with all the CTAs we have, I suggest we leave the CTAs as they are, i.e in uppercase.
- I'am using all the existing font sizes/spacings that we have currently. For e.g the font size of the CTAs are 16px whereas our buttons are set to 14px
- The padding around the widget in the designs is set to 36px but to remain consistent with all our widgets, I've used the default padding we are using for them.
- One major difference between the designs in Figma and what we currently have is regarding the `Dialog` component. Shall I stick to what we currently have or implement the component as per the designs? I suggest we use what we currently have since it'll look like this:
- ![image](https://user-images.githubusercontent.com/31818/141100026-051dc39b-640f-4333-a3f5-01c16f073c04.png)
- compared to ![image](https://user-images.githubusercontent.com/31818/141100141-49967a7a-16d3-4892-998e-9ec1c32edb4c.png) where we also have a different loading state. I'am not sure whether we accounted for those aesthetic changes in the AC or in the estimates.

**Comment by tofumatt** [comment]
> In the Figma designs, the CTAs are normal case ("Connect", "Not interested", "Complete setup") however I remember there were some conversations about them being uppercase like we have right now. To be consistent with all the CTAs we have, I suggest we leave the CTAs as they are, i.e in uppercase.

Sounds good, let's go with the uppercase _styles_. They should be CSS uppercased though, not actually written as text that way 🙂 

> I'am using all the existing font sizes/spacings that we have currently. For e.g the font size of the CTAs are 16px whereas our buttons are set to 14px

Sounds good.

> The padding around the widget in the designs is set to 36px but to remain consistent with all our widgets, I've used the default padding we are using for them.

Sounds good too. In general, we should stick with our existing styles. If the Figma mocks have overall changed margins/paddings we should ignore them and use what we already have defined for our styles. Any "big changes" like that we should do in a separate ticket, repo-wide.

> One major difference between the designs in Figma and what we currently have is regarding the Dialog component. Shall I stick to what we currently have or implement the component as per the designs? I suggest we use what we currently have since it'll look like this:

Yes, stick with the existing dialog. Sorry for the confusion there; we should implement these with the styles we have for now 👍🏻  I think those Figma mocks served as both Unified Dashboard wireframes but also as UI explorations. 😓

**Comment by tofumatt** [comment]
(A separate comment for @felixarntz and @marrrmarrr on the last two points.)

> What should be the URL for the link in "AdSense accounts are subject to review and approval by the Google AdSense team."?

Great question—I'm not sure. @marrrmarrr: what is the right link for the above text? I'd think https://support.google.com/adsense/answer/7402256 would make sense but I'm not sure.

@felixarntz: Should we add a tracking event for users who dismiss the AdSense module/CTA widget? If so, what should the event be named?

**Comment by marrrmarrr** [comment]
>what is the right link for the above text?

I'd suggest https://support.google.com/adsense/answer/9724, as it explicitly states what requirements there are for getting approved. 

Side-question about the [screenshot](https://user-images.githubusercontent.com/90871/135109497-26669e2b-1480-4d4f-9a53-1db7a3c7a8bc.png) from the issue description -- wanted to confirm if this an intermediary version, since the tab which contains AdSense here is still called "revenue" (whereas in the [current design](https://www.figma.com/file/VILRieNSLg2DOVyEgFBkXe/%5BMVP%5D-Unified-Dashboard?node-id=855%3A4304) it's been updated to Monetization as a result of the UXR study we ran), and the chips are missing.

**Comment by felixarntz** [comment]
@marrrmarrr Yes, this is an old screenshot, our codebase already uses only "Monetization".

@tofumatt @asvinb
* The support link should be what @marrrmarrr shared above. I've incorporated it into the ACs.
* I also changed one more detail in the ACs: The widget **should not** be registered in the `AREA_ENTITY_DASHBOARD_MONETIZATION_PRIMARY` area. At this point, there is no widget surfacing any AdSense information in the entity dashboard (see related #2409), so it wouldn't make sense to put this CTA there when after activating AdSense the tab would disappear as well.

Please consider those points for the implementation.

**Comment by felixarntz** [comment]
@asvinb Sorry, I had missed your tracking question from before. That's a great point, let's add an event for when the AdSense module is activated through here, as follows:
* Event name: `activate_module`
* Event category: `{ viewContext }_widget-activation-cta`
* Event label: `adsense`

**Comment by wpdarren** [comment]
### QA Update: ❌ 

During bug bash @eclarke1 noticed that the Complete AdSense activation CTA appears when you only partly complete the connection.  From the conversation on Asana and the bug bash, it was decided that this needs fixing.  Since the issue is linked to the changes in this ticket, we've highlighted the issue here. 

![image](https://user-images.githubusercontent.com/73545194/145247594-f65e2cba-aabd-4198-8d49-3b3d87353379.png)

**Comment by asvinb** [comment]
@wpdarren Follow up PR has been created to fix the issue you mentioned.

**Comment by wpdarren** [comment]
### QA Update: ✅ 

**Verified:**

- New widget created in the Monetization widget area when the unifiedDashboard feature flag is enabled.

![image](https://user-images.githubusercontent.com/73545194/145441879-3f39dff5-bd87-4351-aa35-df3bd19f6945.png)

- The widget renders the AdSense CTA widget as mocked up in the Figma file, with the exception of the known differences 
- When the user clicks on the 'Connect' button they are sent to the AdSense connection screen. 
- When the user clicks on the 'Not interested' link the modal appears. When you click on the 'Yes, Remove' button the widget is closed and the 'Monetization' chip is removed. 
- When the user clicks on the 'Not interested' link the modal appears. When you click on 'Cancel' link the modal closes and the widget still appears. 

<details>
 <summary>Screenshot</summary>

![image](https://user-images.githubusercontent.com/73545194/145441977-5d990dfc-ed58-43f2-80a7-a40b5b68213d.png)

</details>

- If AdSense is activated but not connected, the button says 'Complete setup' The issue reported in this ticket is fixed. No other CTA appears. 

<details>
 <summary>Screenshot</summary>

![image](https://user-images.githubusercontent.com/73545194/145441941-135b79ad-b0ca-4e5b-8c2a-08a1fce7a6a0.png)

</details>


- The AdSense link about account review points to `https://support.google.com/adsense/answer/9724`.
